libs/vkd3d/memory.c

```diff
@@ -659,9 +756,303 @@ void vkd3d_memory_allocator_cleanup(struct vkd3d_memory_allocator *allocator, st
+    VkSemaphoreWaitInfo wait_info;
+    uint64_t old_value, new_value;
+    VkResult vr;
+
+    old_value = vkd3d_atomic_uint64_load_explicit(&clear_queue->last_known_value, vkd3d_memory_order_acquire);
+
+    if (old_value >= wait_value)
+        return true;
+
+    if (timeout)
+    {
+        wait_info.sType = VK_STRUCTURE_TYPE_SEMAPHORE_WAIT_INFO_KHR;
+        wait_info.pNext = NULL;
+        wait_info.flags = 0;
+        wait_info.semaphoreCount = 1;
+        wait_info.pSemaphores = &clear_queue->vk_semaphore;
+        wait_info.pValues = &wait_value;
+
+        vr = VK_CALL(vkWaitSemaphoresKHR(device->vk_device, &wait_info, timeout));
+        new_value = wait_value;
+    }
+    else
+    {
+        vr = VK_CALL(vkGetSemaphoreCounterValueKHR(device->vk_device,
+                clear_queue->vk_semaphore, &new_value));
+    }
+
+    if (vr < 0)
+    {
+        ERR("Failed to wait for timeline semaphore, vr %d.\n", vr);
+        return false;
+    }
+
+    while (new_value > old_value)
+    {
+        uint64_t cur_value = vkd3d_atomic_uint64_compare_exchange(&clear_queue->last_known_value,
+                old_value, new_value, vkd3d_memory_order_release, vkd3d_memory_order_acquire);
+
+        if (cur_value == old_value)
+            break;
+
+        old_value = cur_value;
+    }
+
+    return new_value >= wait_value;
+}
+
+static void vkd3d_memory_allocator_wait_allocation(struct vkd3d_memory_allocator *allocator,
+        struct d3d12_device *device, const struct vkd3d_memory_allocation *allocation)
+{
+    struct vkd3d_memory_clear_queue *clear_queue = &allocator->clear_queue;
+    uint64_t wait_value = allocation->clear_semaphore_value;
+    size_t i;
+
+    /* If the clear semaphore has been signaled to the expected value,
+     * the GPU is already done clearing the allocation, and it cannot
+     * be in the clear queue either, so there is nothing to do. */
+    if (vkd3d_memory_allocator_wait_clear_semaphore(allocator, device, wait_value, 0))
+        return;
+
+    /* If the allocation is still in the queue, the GPU has not started
+     * using it yet so we can remove it from the queue and exit. */
+    pthread_mutex_lock(&clear_queue->mutex);
+
+    for (i = 0; i < clear_queue->allocations_count; i++)
+    {
+        if (clear_queue->allocations[i] == allocation)
+        {
+            clear_queue->allocations[i] = clear_queue->allocations[--clear_queue->allocations_count];
+            clear_queue->num_bytes_pending -= allocation->resource.size;
+            pthread_mutex_unlock(&clear_queue->mutex);
+            return;
+        }
+    }
+
+    /* If this is a chunk and a suballocation from it had been immediately
+     * freed, it is possible that the suballocation got removed from the
+     * clear queue so that the chunk's wait value never gets signaled. Wait
+     * for the last signaled value in that case. */
+    if (wait_value == clear_queue->next_signal_value)
+        wait_value = clear_queue->next_signal_value - 1;
```

Comment:
Where is wait_value used? Below, it's still allocation->clear_semaphore_value

Reply:
Fixed.